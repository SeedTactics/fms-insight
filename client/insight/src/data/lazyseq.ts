import { Vector, HashMap, WithEquality, Option, ToOrderable, Ordering, HashSet, areEqual } from "prelude-ts";

export class LazySeq<T> {
  static ofIterable<T>(iter: Iterable<T>): LazySeq<T> {
    return new LazySeq(iter);
  }

  static ofIterator<T>(f: () => Iterator<T>): LazySeq<T> {
    return new LazySeq<T>({
      [Symbol.iterator]() {
        return f();
      },
    });
  }

  static ofObject<V>(obj: { [k: string]: V }): LazySeq<[string, V]> {
    return LazySeq.ofIterator(function* () {
      for (const k in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, k)) {
          yield [k, obj[k]] as [string, V];
        }
      }
    });
  }

  static ofRange(start: number, end: number, step?: number): LazySeq<number> {
    const s = step || 1;
    return LazySeq.ofIterator(function* () {
      for (let x = start; x < end; x += s) {
        yield x;
      }
    });
  }

  [Symbol.iterator](): Iterator<T> {
    return this.iter[Symbol.iterator]();
  }

  allMatch(f: (x: T) => boolean): boolean {
    for (const x of this.iter) {
      if (!f(x)) {
        return false;
      }
    }
    return true;
  }

  anyMatch(f: (x: T) => boolean): boolean {
    for (const x of this.iter) {
      if (f(x)) {
        return true;
      }
    }
    return false;
  }

  append(x: T): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      yield* iter;
      yield x;
    });
  }

  appendAll(i: Iterable<T>): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      yield* iter;
      yield* i;
    });
  }

  arrangeBy<K>(f: (x: T) => K & WithEquality): Option<HashMap<K, T>> {
    let m = HashMap.empty<K, T>();
    for (const x of this.iter) {
      const k = f(x);
      if (m.containsKey(k)) {
        return Option.none();
      } else {
        m = m.put(k, x);
      }
    }
    return Option.some(m);
  }

  chunk(size: number): LazySeq<ReadonlyArray<T>> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      let chunk: T[] = [];
      for (const x of iter) {
        chunk.push(x);
        if (chunk.length === size) {
          yield chunk;
          chunk = [];
        }
      }
      if (chunk.length > 0) {
        yield chunk;
      }
    });
  }

  contains(v: T & WithEquality): boolean {
    for (const x of this.iter) {
      if (areEqual(x, v)) {
        return true;
      }
    }
    return false;
  }

  drop(n: number): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      let cnt = 0;
      for (const x of iter) {
        if (cnt >= n) {
          yield x;
        } else {
          cnt += 1;
        }
      }
    });
  }

  dropWhile(f: (x: T) => boolean): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      let dropping = true;
      for (const x of iter) {
        if (dropping) {
          if (!f(x)) {
            dropping = false;
            yield x;
          }
        } else {
          yield x;
        }
      }
    });
  }

  isEmpty(): boolean {
    const first = this.iter[Symbol.iterator]().next();
    return first.done === true;
  }

  filter(f: (x: T) => boolean): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      for (const x of iter) {
        if (f(x)) {
          yield x;
        }
      }
    });
  }

  find(f: (v: T) => boolean): Option<T> {
    for (const x of this.iter) {
      if (f(x)) {
        return Option.of(x);
      }
    }
    return Option.none<T>();
  }

  flatMap<S>(f: (x: T) => Iterable<S>): LazySeq<S> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      for (const x of iter) {
        yield* f(x);
      }
    });
  }

  foldLeft<S>(zero: S, f: (soFar: S, cur: T) => S): S {
    let soFar = zero;
    for (const x of this.iter) {
      soFar = f(soFar, x);
    }
    return soFar;
  }

  groupBy<K>(f: (x: T) => K & WithEquality): HashMap<K, Vector<T>> {
    return this.toMap<K & WithEquality, Vector<T>>(
      (x) => [f(x), Vector.of(x)] as [K & WithEquality, Vector<T>],
      (v1, v2) => v1.appendAll(v2)
    );
  }

  head(): Option<T> {
    const first = this.iter[Symbol.iterator]().next();
    if (first.done) {
      return Option.none<T>();
    } else {
      return Option.some(first.value);
    }
  }

  last(): Option<T> {
    let last = Option.none<T>();
    for (const x of this.iter) {
      last = Option.some(x);
    }
    return last;
  }

  length(): number {
    let cnt = 0;
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    for (const _ of this.iter) {
      cnt += 1;
    }
    return cnt;
  }

  map<S>(f: (x: T, idx: number) => S): LazySeq<S> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      let idx = 0;
      for (const x of iter) {
        yield f(x, idx);
        idx += 1;
      }
    });
  }

  mapOption<S>(f: (x: T) => Option<S>): LazySeq<S> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      for (const x of iter) {
        const y = f(x);
        if (y.isSome()) {
          yield y.get();
        }
      }
    });
  }

  maxBy(compare: (v1: T, v2: T) => Ordering): Option<T> {
    return this.reduce((v1, v2) => (compare(v1, v2) < 0 ? v2 : v1));
  }

  maxOn(getOrderable: ToOrderable<T>): Option<T> {
    let ret = Option.none<T>();
    let maxVal: number | string | boolean | undefined;
    for (const x of this.iter) {
      if (ret.isNone()) {
        ret = Option.of(x);
        maxVal = getOrderable(x);
      } else {
        const curVal = getOrderable(x);
        if (!maxVal || maxVal < curVal) {
          ret = Option.of(x);
          maxVal = curVal;
        }
      }
    }
    return ret;
  }

  minBy(compare: (v1: T, v2: T) => Ordering): Option<T> {
    return this.reduce((v1, v2) => (compare(v1, v2) > 0 ? v2 : v1));
  }

  minOn(getOrderable: ToOrderable<T>): Option<T> {
    let ret = Option.none<T>();
    let minVal: number | string | boolean | undefined;
    for (const x of this.iter) {
      if (ret.isNone()) {
        ret = Option.of(x);
        minVal = getOrderable(x);
      } else {
        const curVal = getOrderable(x);
        if (!minVal || minVal > curVal) {
          ret = Option.of(x);
          minVal = curVal;
        }
      }
    }
    return ret;
  }

  prepend(x: T): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      yield x;
      yield* iter;
    });
  }

  prependAll(i: Iterable<T>): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      yield* i;
      yield* iter;
    });
  }

  reduce(f: (x1: T, x2: T) => T): Option<T> {
    let ret = Option.none<T>();
    for (const x of this.iter) {
      if (ret.isNone()) {
        ret = Option.of(x);
      } else {
        ret = Option.of(f(ret.get(), x));
      }
    }
    return ret;
  }

  single(): Option<T> {
    const iter = this.iter[Symbol.iterator]();
    const first = iter.next();
    if (first.done) {
      // empty
      return Option.none<T>();
    }
    const next = iter.next();
    if (next.done) {
      // singleton
      return Option.some(first.value);
    } else {
      // more than one
      return Option.none<T>();
    }
  }

  sortBy(compare: (v1: T, v2: T) => Ordering): LazySeq<T> {
    const arr = this.toArray().sort(compare);
    return LazySeq.ofIterable(arr);
  }

  sortOn(...getKeys: Array<ToOrderable<T> | { desc: ToOrderable<T> }>): LazySeq<T> {
    return this.sortBy((x, y) => {
      for (const getKey of getKeys) {
        if ((<any>getKey).desc) {
          const a = (<ToOrderable<T>>(<any>getKey).desc)(x);
          const b = (<ToOrderable<T>>(<any>getKey).desc)(y);
          if (a === b) {
            continue;
          }
          return a < b ? 1 : -1;
        } else {
          const a = (<ToOrderable<T>>getKey)(x);
          const b = (<ToOrderable<T>>getKey)(y);
          if (a === b) {
            continue;
          }
          return a < b ? -1 : 1;
        }
      }
      return 0;
    });
  }

  sumOn(getNumber: (v: T) => number): number {
    let sum = 0;
    for (const x of this.iter) {
      sum += getNumber(x);
    }
    return sum;
  }

  tail(): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      let seenFirst = false;
      for (const x of iter) {
        if (seenFirst) {
          yield x;
        } else {
          seenFirst = true;
        }
      }
    });
  }

  take(n: number): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      let cnt = 0;
      for (const x of iter) {
        if (cnt >= n) {
          return;
        }
        yield x;
        cnt += 1;
      }
    });
  }

  takeWhile(f: (x: T) => boolean): LazySeq<T> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      for (const x of iter) {
        if (f(x)) {
          yield x;
        } else {
          return;
        }
      }
    });
  }

  toArray(): Array<T> {
    return Array.from(this.iter);
  }

  toMap<K, S>(f: (x: T) => [K & WithEquality, S], merge: (v1: S, v2: S) => S): HashMap<K, S> {
    let m = HashMap.empty<K, S>();
    for (const x of this.iter) {
      const [k, s] = f(x);
      m = m.putWithMerge(k, s, merge);
    }
    return m;
  }

  toSet<S>(converter: (x: T) => S & WithEquality): HashSet<S> {
    const iter = this.iter;
    return HashSet.ofIterable({
      *[Symbol.iterator]() {
        for (const x of iter) {
          yield converter(x);
        }
      },
    });
  }

  toVector(): Vector<T> {
    return Vector.ofIterable(this.iter);
  }

  transform<U>(f: (s: LazySeq<T>) => U): U {
    return f(this);
  }

  zip<S>(other: Iterable<S>): LazySeq<[T, S]> {
    const iter = this.iter;
    return LazySeq.ofIterator(function* () {
      const i1 = iter[Symbol.iterator]();
      const i2 = other[Symbol.iterator]();
      while (true) {
        const n1 = i1.next();
        const n2 = i2.next();
        if (n1.done || n2.done) {
          return;
        }
        yield [n1.value, n2.value] as [T, S];
      }
    });
  }

  private constructor(private iter: Iterable<T>) {}
}
