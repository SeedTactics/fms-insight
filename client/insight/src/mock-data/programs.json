[
  {
    "MachineGroupName": "Machine",
    "ProgramName": "100",
    "CellControllerProgramName": "100",
    "Revision": null,
    "Comment": "program 100 comment"
  },
  {
    "MachineGroupName": "Machine",
    "ProgramName": "200",
    "CellControllerProgramName": "200",
    "Revision": null,
    "Comment": "comment for 200"
  },
  {
    "MachineGroupName": "Machine",
    "ProgramName": "300",
    "Revision": null,
    "CellControllerProgramName": "300"
  },
  {
    "MachineGroupName": "Machine",
    "ProgramName": "400",
    "CellControllerProgramName": "400",
    "Revision": null
  }
]
