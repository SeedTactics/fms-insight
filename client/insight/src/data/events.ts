/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import { addDays, addMonths, startOfMonth } from "date-fns";
import { PledgeStatus, Pledge, ActionBeforeMiddleware, PledgeToPromise } from "../store/middleware";
import { Vector } from "prelude-ts";
import { LazySeq } from "./lazyseq";

import * as api from "./api";
import * as cycles from "./events.cycles";
import * as matsummary from "./events.matsummary";
import * as simuse from "./events.simuse";
import * as inspection from "./events.inspection";
import { JobsBackend, LogBackend } from "./backend";

export enum AnalysisPeriod {
  Last30Days = "Last_30_Days",
  SpecificMonth = "Specific_Month",
}

export interface Last30Days {
  readonly latest_log_counter: number | undefined;
  readonly most_recent_10_events: Vector<Readonly<api.ILogEntry>>;

  readonly latest_scheduleId: string | undefined;

  readonly cycles: cycles.CycleState;
  readonly sim_use: simuse.SimUseState;
  readonly inspection: inspection.InspectionState;

  readonly mat_summary: matsummary.MatSummaryState;
}

export interface AnalysisMonth {
  readonly cycles: cycles.CycleState;
  readonly sim_use: simuse.SimUseState;
  readonly inspection: inspection.InspectionState;
}

const emptyAnalysisMonth: AnalysisMonth = {
  cycles: cycles.initial,
  sim_use: simuse.initial,
  inspection: inspection.initial,
};

export interface State {
  readonly loading_log_entries: boolean;
  readonly loading_job_history: boolean;
  readonly loading_error?: Error;

  readonly analysis_period: AnalysisPeriod;
  readonly analysis_period_month: Date;
  readonly loading_analysis_month_log: boolean;
  readonly loading_analysis_month_jobs: boolean;

  readonly last30: Last30Days;
  readonly selected_month: AnalysisMonth;
}

export const initial: State = {
  loading_log_entries: false,
  loading_job_history: false,
  analysis_period: AnalysisPeriod.Last30Days,
  analysis_period_month: startOfMonth(new Date()),
  loading_analysis_month_log: false,
  loading_analysis_month_jobs: false,

  last30: {
    latest_log_counter: undefined,
    latest_scheduleId: undefined,
    most_recent_10_events: Vector.empty(),
    cycles: cycles.initial,
    mat_summary: matsummary.initial,
    sim_use: simuse.initial,
    inspection: inspection.initial,
  },

  selected_month: emptyAnalysisMonth,
};

export enum ActionType {
  SetAnalysisLast30Days = "Events_SetAnalysisLast30Days",
  SetAnalysisMonth = "Events_SetAnalysisMonth",
  LoadRecentLogEntries = "Events_LoadRecentLogEntries",
  LoadRecentJobHistory = "Events_LoadRecentJobHistory",
  LoadSpecificMonthLogEntries = "Events_LoadSpecificMonthLogEntries",
  LoadSpecificMonthJobHistory = "Events_LoadSpecificMonthJobHistory",
  ReceiveNewLogEntries = "Events_NewLogEntries",
  ReceiveNewJobs = "Events_ReceiveNewJobs",
}

export type Action =
  | { type: ActionType.SetAnalysisLast30Days }
  | { type: ActionType.SetAnalysisMonth; month: Date }
  | {
      type: ActionType.LoadRecentLogEntries;
      now: Date;
      pledge: Pledge<ReadonlyArray<Readonly<api.ILogEntry>>>;
    }
  | {
      type: ActionType.LoadRecentJobHistory;
      now: Date;
      pledge: Pledge<Readonly<api.IHistoricData>>;
    }
  | {
      type: ActionType.LoadSpecificMonthLogEntries;
      month: Date;
      pledge: Pledge<ReadonlyArray<Readonly<api.ILogEntry>>>;
    }
  | {
      type: ActionType.LoadSpecificMonthJobHistory;
      month: Date;
      pledge: Pledge<Readonly<api.IHistoricData>>;
    }
  | {
      type: ActionType.ReceiveNewLogEntries;
      now: Date;
      events: ReadonlyArray<Readonly<api.ILogEntry>>;
    }
  | {
      type: ActionType.ReceiveNewJobs;
      now: Date;
      jobs: Readonly<api.IHistoricData>;
    };

type ABF = ActionBeforeMiddleware<Action>;

export function loadLast30Days(): ABF {
  const now = new Date();
  const thirtyDaysAgo = addDays(now, -30);
  return [
    {
      type: ActionType.LoadRecentLogEntries,
      now: now,
      pledge: LogBackend.get(thirtyDaysAgo, now),
    },
    {
      type: ActionType.LoadRecentJobHistory,
      now: now,
      pledge: JobsBackend.history(thirtyDaysAgo, now),
    },
  ];
}

export function refreshLogEntries(lastCounter: number): ABF {
  const now = new Date();
  return {
    type: ActionType.LoadRecentLogEntries,
    now: now,
    pledge: LogBackend.recent(lastCounter),
  };
}

export function receiveNewEvents(events: ReadonlyArray<Readonly<api.ILogEntry>>): ABF {
  return {
    type: ActionType.ReceiveNewLogEntries,
    now: new Date(),
    events,
  };
}

export function receiveNewJobs(newJobs: Readonly<api.INewJobs>): ABF {
  const jobs: { [key: string]: api.JobPlan } = {};
  newJobs.jobs.forEach((j) => {
    jobs[j.unique] = j;
  });
  return {
    type: ActionType.ReceiveNewJobs,
    now: new Date(),
    jobs: {
      jobs: jobs,
      stationUse: newJobs.stationUse || [],
    },
  };
}

export function analyzeLast30Days(): ABF {
  return { type: ActionType.SetAnalysisLast30Days };
}

export function analyzeSpecificMonth(month: Date): ReadonlyArray<PledgeToPromise<Action>> {
  const startOfNextMonth = addMonths(month, 1);
  return [
    {
      type: ActionType.LoadSpecificMonthLogEntries,
      month: month,
      pledge: LogBackend.get(month, startOfNextMonth),
    },
    {
      type: ActionType.LoadSpecificMonthJobHistory,
      month: month,
      pledge: JobsBackend.history(month, startOfNextMonth),
    },
  ];
}

export function setAnalysisMonth(month: Date): ABF {
  return {
    type: ActionType.SetAnalysisMonth,
    month: month,
  };
}

function safeAssign<T extends R, R>(o: T, n: R): T {
  let allMatch = true;
  for (const k of Object.keys(n)) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    if ((o as any)[k] !== (n as any)[k]) {
      allMatch = false;
      break;
    }
  }
  if (allMatch) {
    return o;
  } else {
    return Object.assign({}, o, n);
  }
}

function processRecentLogEntries(now: Date, evts: ReadonlyArray<Readonly<api.ILogEntry>>, s: Last30Days): Last30Days {
  const thirtyDaysAgo = addDays(now, -30);
  const initialLoad = s.latest_log_counter === undefined;
  let lastCounter = s.latest_log_counter;
  const lastNewEvent = LazySeq.ofIterable(evts).maxOn((e) => e.counter);
  let last10Evts = s.most_recent_10_events;
  if (lastNewEvent.isSome()) {
    if (lastCounter === undefined || lastCounter < lastNewEvent.get().counter) {
      lastCounter = lastNewEvent.get().counter;
    }
    const lastNew10 = evts.slice(-10);
    last10Evts = last10Evts.appendAll(lastNew10).reverse().take(10).reverse();
  }
  return safeAssign(s, {
    latest_log_counter: lastCounter,
    cycles: cycles.process_events(
      { type: cycles.ExpireOldDataType.ExpireEarlierThan, d: thirtyDaysAgo },
      evts,
      initialLoad,
      s.cycles
    ),
    most_recent_10_events: last10Evts,
    mat_summary: matsummary.process_events(now, evts, s.mat_summary),
    inspection: inspection.process_events(
      { type: cycles.ExpireOldDataType.ExpireEarlierThan, d: thirtyDaysAgo },
      evts,
      undefined,
      s.inspection
    ),
  });
}

function processSpecificMonthLogEntries(evts: ReadonlyArray<Readonly<api.ILogEntry>>, s: AnalysisMonth): AnalysisMonth {
  return safeAssign(s, {
    cycles: cycles.process_events(
      { type: cycles.ExpireOldDataType.NoExpire },
      evts,
      true, // initial load is true
      s.cycles
    ),
    inspection: inspection.process_events({ type: cycles.ExpireOldDataType.NoExpire }, evts, undefined, s.inspection),
  });
}

function processRecentJobs(now: Date, jobs: Readonly<api.IHistoricData>, s: Last30Days): Last30Days {
  const thirtyDaysAgo = addDays(now, -30);
  let latestSchId = s.latest_scheduleId;
  const largestSchId = LazySeq.ofObject(jobs.jobs)
    .maxOn(([_, j]) => j.scheduleId || "")
    .mapNullable(([_, j]) => j.scheduleId);
  if (largestSchId.isSome()) {
    if (latestSchId === undefined || latestSchId < largestSchId.get()) {
      latestSchId = largestSchId.get();
    }
  }

  return safeAssign(s, {
    latest_scheduleId: latestSchId,
    sim_use: simuse.process_sim_use(
      { type: cycles.ExpireOldDataType.ExpireEarlierThan, d: thirtyDaysAgo },
      jobs,
      s.sim_use
    ),
  });
}

function processSpecificMonthJobs(jobs: Readonly<api.IHistoricData>, s: AnalysisMonth): AnalysisMonth {
  return safeAssign(s, {
    sim_use: simuse.process_sim_use({ type: cycles.ExpireOldDataType.NoExpire }, jobs, s.sim_use),
  });
}

export function reducer(s: State | undefined, a: Action): State {
  if (s === undefined) {
    return initial;
  }
  switch (a.type) {
    case ActionType.LoadRecentLogEntries:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return { ...s, loading_log_entries: true, loading_error: undefined };
        case PledgeStatus.Completed:
          return {
            ...s,
            last30: processRecentLogEntries(a.now, a.pledge.result, s.last30),
            loading_log_entries: false,
          };
        case PledgeStatus.Error:
          return {
            ...s,
            loading_log_entries: false,
            loading_error: a.pledge.error,
          };
        default:
          return s;
      }

    case ActionType.ReceiveNewLogEntries:
      return {
        ...s,
        last30: processRecentLogEntries(a.now, a.events, s.last30),
      };

    case ActionType.LoadRecentJobHistory:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return { ...s, loading_job_history: true, loading_error: undefined };
        case PledgeStatus.Completed:
          return {
            ...s,
            last30: processRecentJobs(a.now, a.pledge.result, s.last30),
            loading_job_history: false,
          };
        case PledgeStatus.Error:
          return {
            ...s,
            loading_job_history: false,
            loading_error: a.pledge.error,
          };
        default:
          return s;
      }

    case ActionType.ReceiveNewJobs:
      return {
        ...s,
        last30: processRecentJobs(a.now, a.jobs, s.last30),
      };

    case ActionType.SetAnalysisLast30Days:
      return {
        ...s,
        analysis_period: AnalysisPeriod.Last30Days,
        selected_month: emptyAnalysisMonth,
      };

    case ActionType.LoadSpecificMonthLogEntries:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return {
            ...s,
            analysis_period: AnalysisPeriod.SpecificMonth,
            analysis_period_month: a.month,
            loading_error: undefined,
            loading_analysis_month_log: true,
            selected_month: emptyAnalysisMonth,
          };

        case PledgeStatus.Completed:
          return {
            ...s,
            loading_analysis_month_log: false,
            selected_month: processSpecificMonthLogEntries(a.pledge.result, s.selected_month),
          };

        case PledgeStatus.Error:
          return {
            ...s,
            loading_analysis_month_log: false,
            loading_error: a.pledge.error,
          };

        default:
          return s;
      }

    case ActionType.LoadSpecificMonthJobHistory:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return {
            ...s,
            loading_analysis_month_jobs: true,
          };

        case PledgeStatus.Completed:
          return {
            ...s,
            loading_analysis_month_jobs: false,
            selected_month: processSpecificMonthJobs(a.pledge.result, s.selected_month),
          };

        case PledgeStatus.Error:
          return {
            ...s,
            loading_analysis_month_jobs: false,
            loading_error: a.pledge.error,
          };

        default:
          return s;
      }

    case ActionType.SetAnalysisMonth:
      return { ...s, analysis_period_month: a.month };

    default:
      return s;
  }
}
