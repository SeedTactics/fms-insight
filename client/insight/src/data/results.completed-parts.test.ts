/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { addDays, addHours, differenceInMinutes, addMinutes, differenceInSeconds } from "date-fns";

import { PledgeStatus } from "../store/middleware";
import * as events from "./events";
import { fakeCycle } from "./events.fake";
import { ILogEntry } from "./api";
import {
  binCyclesByDayAndPart,
  buildCompletedPartsTable,
  buildCompletedPartSeries,
  buildCompletedPartsHeatmapTable,
} from "./results.completed-parts";
import { loadMockData } from "../mock-data/load";
import { LazySeq } from "./lazyseq";

it("bins actual cycles by day", () => {
  const now = new Date(2018, 2, 5); // midnight in local time

  const evts = ([] as ILogEntry[]).concat(
    fakeCycle(now, 30),
    fakeCycle(addHours(now, -3), 20),
    fakeCycle(addHours(now, -15), 15)
  );
  const st = events.reducer(events.initial, {
    type: events.ActionType.LoadRecentLogEntries,
    now: addDays(now, 1),
    pledge: {
      status: PledgeStatus.Completed,
      result: evts,
    },
  });

  let byDayAndPart = binCyclesByDayAndPart(st.last30.cycles.part_cycles);

  const points = LazySeq.ofIterable(byDayAndPart)
    .map(([dayAndPart, val]) => ({
      x: dayAndPart.day,
      y: dayAndPart.part,
      label: "Unused",
      count: val.count,
      activeMachineMins: val.activeMachineMins,
    }))
    .toArray();

  const heattable = document.createElement("div");
  heattable.innerHTML = buildCompletedPartsHeatmapTable(points);
  expect(heattable).toMatchSnapshot("heatmap clipboard table");

  // convert to chicago time because snapshot includes date and time in UTC when formatting the byDayAndPart snapshot
  const nowChicago = new Date(Date.UTC(2018, 2, 5, 6, 0, 0)); // America/Chicago time
  const minOffset = differenceInMinutes(nowChicago, now);
  byDayAndPart = byDayAndPart.map((dayAndPart, val) => [dayAndPart.adjustDay((d) => addMinutes(d, minOffset)), val]);

  expect(byDayAndPart).toMatchSnapshot("cycles binned by day and part");
});

it("build completed series", async () => {
  const jan18 = new Date(Date.UTC(2018, 0, 1, 0, 0, 0));
  const today = new Date(2018, 7, 6);
  const todayChicago = new Date(Date.UTC(2018, 7, 6, 5, 0, 0)); // America/Chicago time during DST
  const zoneOffset = differenceInMinutes(todayChicago, today);
  const offsetSeconds = differenceInSeconds(addDays(today, -28), jan18);
  const data = loadMockData(offsetSeconds);
  const evts = await data.events;
  const st = events.reducer(events.initial, {
    type: events.ActionType.LoadRecentLogEntries,
    now: today,
    pledge: {
      status: PledgeStatus.Completed,
      result: evts,
    },
  });

  const origSeries = buildCompletedPartSeries(
    addDays(today, -6),
    today,
    st.last30.cycles.part_cycles,
    st.last30.sim_use.production
  );
  const adjustedSeries = origSeries.map((s) => ({
    ...s,
    days: s.days.map((d) => ({ ...d, day: addMinutes(d.day, zoneOffset) })),
  }));
  expect(adjustedSeries).toMatchSnapshot("completed part series");

  const table = document.createElement("div");
  table.innerHTML = buildCompletedPartsTable(origSeries);
  expect(table).toMatchSnapshot("clipboard table");
});
