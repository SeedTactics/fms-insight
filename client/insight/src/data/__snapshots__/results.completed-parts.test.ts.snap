// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bins actual cycles by day > cycles binned by day and part 1`] = `
Map {
  "{day: 2018-03-04T06:00:00.000Z}, part: partC-2}" => {
    "activeMachineMins": 20,
    "count": 1,
  },
  "{day: 2018-03-04T06:00:00.000Z}, part: partz-3}" => {
    "activeMachineMins": 15,
    "count": 1,
  },
}
`;

exports[`bins actual cycles by day > heatmap clipboard table 1`] = `
<div>
  <table>
    

    <thead>
      <tr>
        <th>
          Part
        </th>
        <th>
          Expected Cycle Mins
        </th>
        <th>
          Sun Mar 04 2018 Completed
        </th>
        <th>
          Sun Mar 04 2018 Std. Hours
        </th>
      </tr>
    </thead>
    

    <tbody>
      

      <tr>
        <th>
          partC-2
        </th>
        <th>
          20.0
        </th>
        <td>
          1
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <th>
          partz-3
        </th>
        <th>
          15.0
        </th>
        <td>
          1
        </td>
        <td>
          0.3
        </td>
      </tr>
      

    </tbody>
    

  </table>
</div>
`;
